src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -379,15 +481,165 @@ internal static string GetStringForPartUri(Uri partUri)
+            if (partUri != null)
+                packUri = new Uri(packUri, partUri);
+
+            // Step 7 - Append fragment if present
+            if (fragment != null)
+                packUri = new Uri(String.Concat(packUri.GetComponents(UriComponents.AbsoluteUri, UriFormat.UriEscaped), fragment));
+            
+            // We want to ensure that internal content of resulting Uri has canonical form
+            // i.e.  result.OrignalString would appear as perfectly formatted Uri string 
+            // so we roundtrip the result.
+            
+            return new Uri(packUri.GetComponents(UriComponents.AbsoluteUri, UriFormat.UriEscaped));
+        }
+
+        private static void EnsurePackSchemeRegistered()
+        {
+            if (!UriParser.IsKnownScheme(UriSchemePack))
+            {
+                // Indicate that we want a default hierarchical parser with a registry based authority
+                UriParser.Register(new GenericUriParser(GenericUriParserOptions.GenericAuthority), UriSchemePack, -1);
+            }
+        }
+
+        /// <summary>
+        /// This method is used to validate the package uri
+        /// </summary>
+        /// <param name="packageUri"></param>
+        /// <returns></returns>
+        private static Uri ValidatePackageUri(Uri packageUri)
+        {
+            if (packageUri == null)
+                throw new ArgumentNullException(nameof(packageUri));
+
+            if (!packageUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            return packageUri;
+        }
+
+        //validates is a given uri has pack:// scheme
+        private static Uri ValidatePackUri(Uri packUri)
+        {
+            if (packUri == null)
+                throw new ArgumentNullException(nameof(packUri));
+
+            if (!packUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            if (packUri.Scheme != PackUriHelper.UriSchemePack)
+                throw new ArgumentException(SR.UriShouldBePackScheme);
+
+            return packUri;
+        }
+
+        /// <summary>
+        /// Escapes -  %', '@', ',', '?' in the package URI 
+        /// This method modifies the string in a culture safe and case safe manner. 
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        private static string EscapeSpecialCharacters(string path)
+        {            
+            string characterString;
+
+            // Escaping for the following - '%'; '@'; ',' and '?'
+            // !!Important!! - The order is important - The '%' sign should be escaped first.
+            // This is currently enforced by the order of characters in the _specialCharacters array
+            foreach (char c in s_specialCharacterChars)
+            {
+                characterString = c.ToString();
+                if (path.Contains(characterString))
```

Comment:
This ToString should not be necessary, at least not outside of the if block. String.Contains has an overload that takes a char.

Reply:
And it really shouldn't ever be necessary, since this is a fixed list of chars for which we could just store the strings.